### PR TITLE
Software Testing Part 1 - Introduction to Unit Testing

### PR DESCRIPTION
**Pull request by Quadrisheriff**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Quadrisheriff** [comment]
This fixes #718

**Comment by hectorkambow** [comment]
Plagiarism check done

File: articles/software-testing-part1/index.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+def test_find_factorial_of_negative_numbers():
+    num = -1
+    assert find_factorial(num) == ValueError
+
+def test_find_factorial():
+    num = 5
+    assert find_factorial(num) == 120
+```
+Run your tests to confirm if your code unit is working as expected.
```
Comment (adrianmurage) [inline articles/software-testing-part1/index.md:76]: the test results I am getting for this:
``` bash

num = -941

    def  find_factorial(num):
>       if num == 1:
E       RecursionError: maximum recursion depth exceeded in comparison

factorial.py:2: RecursionError
_______________________ test_find_factorial ________________________

    def test_find_factorial():
        num = 5
>       assert find_factorial(num) == 120
E       AssertionError: assert <class 'ValueError'> == 120
E        +  where <class 'ValueError'> = find_factorial(5)

_test.py:20: AssertionError
===================== short test summary info ======================
FAILED _test.py::test_find_factorial_of_negative_numbers - Recurs...
FAILED _test.py::test_find_factorial - AssertionError: assert <cl...
=================== 2 failed, 2 passed in 1.29s ====================
```

File: articles/software-testing-part1/index.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    else:
+        return num * find_factorial(num - 1)
+```
+
+To perform unit testing, you have to test every output of your code unit with their expected return value to confirm if they adhere to their contract. 
+
+```python
+# _test.py
+
+Import pytest
```
[the remark this answers is not in the document]
Comment (adrianmurage) [reply articles/software-testing-part1/index.md:58]: If the function being tested is in a separate file 
you would need to import the function

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@Quadrisheriff 
Great article 
But there's one major flaw in the article. This being an introductory article you need to cover setup and go through the testing process step by step.

1. Setup a virtual env
2. Install pytest
3. Folder structure (it is unclear whether or not the function and tests are located in the same file. Make that clear)
4. Give the command needed to run the tests

**Comment by adrianmurage** [comment]
@Quadrisheriff 
I'm guessing the initial problem is with the factorial function.
Have you tried running it to see whether it actually works?

**Comment by Quadrisheriff** [comment]
> @Quadrisheriff
> Great article
> But there's one major flaw in the article. This being an introductory article you need to cover setup and go through the testing process step by step.
> 
> 1. Setup a virtual env
> 2. Install pytest
> 3. Folder structure (it is unclear whether or not the function and tests are located in the same file. Make that clear)
> 4. Give the command needed to run the tests

Hi, I actually thought about adding all of this when I was writing the article but decided against it. I want the article to be focused mainly on running unit tests which you can do in any language with plenty of frameworks.

Adding pytests installation and setup may drag the article focus away from unit testing to a pytest tutorial(which is a separate long topic on its own) which I don't want. I already added a link to the pytest tutorial where the user can learn more about pytest.

**Comment by Quadrisheriff** [comment]
I have made some changes to the code samples, I think you should rerun the tests now. I am not getting any errors here.

**Comment by adrianmurage** [comment]
> 
> 
> > @Quadrisheriff
> > Great article
> > But there's one major flaw in the article. This being an introductory article you need to cover setup and go through the testing process step by step.
> > 
> > 1. Setup a virtual env
> > 2. Install pytest
> > 3. Folder structure (it is unclear whether or not the function and tests are located in the same file. Make that clear)
> > 4. Give the command needed to run the tests
> 
> Hi, I actually thought about adding all of this when I was writing the article but decided against it. I want the article to be focused mainly on running unit tests which you can do in any language with plenty of frameworks.
> 
> Adding pytests installation and setup may drag the article focus away from unit testing to a pytest tutorial(which is a separate long topic on its own) which I don't want. I already added a link to the pytest tutorial where the user can learn more about pytest. I just want to show the ready an example of how to run unit tests and nothing more.

My question to you is a simple one.
If I don't know how to set up the test environment, will I be able to run the tests?

consider adding the setup 
In my opinion, it is pretty straight forward.

1. create a project folder
`mkdir testing`
2. create a virtual environment and activate it
`python3 -m venv virtual && source virtual/bin/activate`
3. create the two necessary files
` touch _test.py factorial.py`

And that's all you would need to add to help the reader get setup.

**Comment by Quadrisheriff** [comment]
> > > @Quadrisheriff
> > > Great article
> > > But there's one major flaw in the article. This being an introductory article you need to cover setup and go through the testing process step by step.
> > > 
> > > 1. Setup a virtual env
> > > 2. Install pytest
> > > 3. Folder structure (it is unclear whether or not the function and tests are located in the same file. Make that clear)
> > > 4. Give the command needed to run the tests
> > 
> > 
> > Hi, I actually thought about adding all of this when I was writing the article but decided against it. I want the article to be focused mainly on running unit tests which you can do in any language with plenty of frameworks.
> > Adding pytests installation and setup may drag the article focus away from unit testing to a pytest tutorial(which is a separate long topic on its own) which I don't want. I already added a link to the pytest tutorial where the user can learn more about pytest. I just want to show the ready an example of how to run unit tests and nothing more.
> 
> My question to you is a simple one.
> If I don't know how to set up the test environment, will I be able to run the tests?
> 
> consider adding the setup
> In my opinion, it is pretty straight forward.
> 
> 1. create a project folder
>    `mkdir testing`
> 2. create a virtual environment and activate it
>    `python3 -m venv virtual && source virtual/bin/activate`
> 3. create the two necessary files
>    ` touch _test.py factorial.py`
> 
> And that's all you would need to add to help the reader get setup.

**Review by adrianmurage — COMMENTED** [review]
As pointed out earlier, your code for the factorial is incorrect. 
Kindly run double check it.
Below is a version that passes the unit tests
```python
def  find_factorial(num):
    factorial = 1

    if num == 1:
        return factorial
    elif  num == 0:
        return factorial
    elif num < 0:
        return  ValueError
    else:
        for i in range(1, (num + 1)):
            factorial = factorial * i
        return factorial
```

**Comment by Quadrisheriff** [comment]
> As pointed out earlier, your code for the factorial is incorrect.
> Kindly run double check it.
> Below is a version that passes the unit tests
> 
> ```python
> def  find_factorial(num):
>     factorial = 1
> 
>     if num == 1:
>         return factorial
>     elif  num == 0:
>         return factorial
>     elif num < 0:
>         return  ValueError
>     else:
>         for i in range(1, (num + 1)):
>             factorial = factorial * i
>         return factorial
> ```

I have gone through code samples. Mine is working well and still passing all my tests. I am actually very confused now, but I will change the code to yours because yours is passing here too. 

I have already added the setup part to the article. Thanks a lot for your advice tho, I hope I haven't sounded rude anywhere in my replies.

**Comment by Quadrisheriff** [comment]
Please check my article now.

**Review by adrianmurage — APPROVED** [review]

**Comment by adrianmurage** [comment]
> 
> 
> > As pointed out earlier, your code for the factorial is incorrect.
> > Kindly run double check it.
> > Below is a version that passes the unit tests
> > ```python
> > def  find_factorial(num):
> >     factorial = 1
> > 
> >     if num == 1:
> >         return factorial
> >     elif  num == 0:
> >         return factorial
> >     elif num < 0:
> >         return  ValueError
> >     else:
> >         for i in range(1, (num + 1)):
> >             factorial = factorial * i
> >         return factorial
> > ```
> 
> I have gone through code samples. Mine is working well and still passing all my tests. I am actually very confused now, but I will change the code to yours because yours is passing here too.
> 
> I have already added the setup part to the article. Thanks a lot for your advice tho, I hope I haven't sounded rude anywhere in my replies.

The aim of the articles we write here at section is to help other engineers in their learning, that said your guides and tutorials should always work for whoever is following through with your article.
My advice would be, whenever you're teaching something through your article try it out for yourself.
Create an isolated environment (a virtual env or a separate user on your Linux instance) and follow through with your article. This will help you point out things you might have overlooked when writing your article. It will also give you the confidence to say that your guide works for the reader.

As for sounding rude, no you haven't. 
Your article is a good one and will be a valuable contribution to EngEd, keep up the good work and take note of the things I pointed out and follow them

**Comment by adrianmurage** [comment]
@hectorkambow 
This article is ready for final review

**Pull request merged**